GitHub proxy part 6.5: tsh git ssh/clone/config

greedy52 opened the pull request:
related:
- #48762
- #49980

examples:
```
$ tsh git config
The current Git directory is configured with Teleport for GitHub organization "goteleport-core-test".

$ tsh git config reset
Teleport configuration removed.

$ tsh git config
The current Git directory is not configured with Teleport.
Run 'tsh git config update' to configure it.

$ tsh git config update
Teleport configuration added.
The current Git directory is configured with Teleport for GitHub organization "goteleport-core-test".
```

greedy52 committed:
GitHub proxy part 6.5: tsh git ssh/clone/config

greedy52 commented on go.mod:
```diff
@@ -100,6 +100,7 @@ require (
 	github.com/fxamacker/cbor/v2 v2.7.0
 	github.com/ghodss/yaml v1.0.0
 	github.com/gizak/termui/v3 v3.1.0
+	github.com/go-git/go-git/v5 v5.12.0
```
`go-git` will be used for the `teleport` as well.

local mac build `tsh` increased about 200k (out of 106MB) after importing it.

Tener commented on tool/tsh/common/git.go:
```diff
@@ -19,24 +19,119 @@
 	cmds.login.Hidden()
 	cmds.list.Hidden()
+	cmds.config.Hidden()
+	cmds.clone.Hidden()
```
Should we hide `cmds.ssh` too?

greedy52 replied:
`cmds.ssh` is always hidden

Tener commented on tool/tsh/common/git.go:
```diff
@@ -19,24 +19,119 @@
+	return g.Host == "github.com"
+}
+
+// owner returns the first part of the path. If the path does not have an owner,
+// an empty string is returned.
+//
+// For GitHub, owner is either the user or the organization that owns the repo.
```
Some examples would be nice.

Tener commented on tool/tsh/common/git.go:
```diff
@@ -19,24 +19,119 @@
+//
+// For GitHub, owner is either the user or the organization that owns the repo.
+func (g gitSSHURL) owner() string {
+	owner, _, ok := strings.Cut(strings.TrimPrefix(g.Path, "/"), "/")
```
How reliable is this? I'm worried about unintended consequences if `g.Path` is shaped differently from our expectations. Perhaps we can ensure some well-known prefix is present?

greedy52 replied:
i would say it's quite reliable but at the end of the day it depends on how good is `go-git` lib's parsing. I have added a comment on the `/` prefix. It only happens for url like path where it has a preceding `/`. SCP-like paths do not have it.

Tener commented on tool/tsh/common/git.go:
```diff
@@ -19,24 +19,119 @@
 }
+
+type gitSSHURL struct {
+	*transport.Endpoint
```
nit: my personal preference is to name each field explicitly to avoid confusion, e.g.:

```suggestion
	Endpoint *transport.Endpoint
```

Again, just my personal view, feel free to ignore.

greedy52 replied:
ended up like this:
```go
type gitSSHURL transport.Endpoint
```
there is no need for other attributes (yet).

Tener commented on tool/tsh/common/git_clone.go:
```diff
@@ -0,0 +1,71 @@
+		return trace.Wrap(err)
+	}
+	if !u.isGitHub() {
+		return trace.BadParameter("not a GitHub repository")
```
May lead to better UX if we include the faulty URL in the error message.

Tener commented on tool/tsh/common/git_config.go:
```diff
@@ -0,0 +1,178 @@
+
+// gitConfigCommand implements `tsh git config`.
+//
+// This command internally executes `git` commands like `git config xxx`. You
+// can generally assume the user has `git` binary installed (otherwise there is
+// no point using the `git` proxy feature). An alternative is to use the
```
The dependency on `git` binaries isn't completely obvious. I think it is worth giving a clear error message designed to catch this issue.

The other day I was installing VSCode on a fresh machine and it clearly shown a custom error message designed to catch exactly this: lack of `git` binaries. I really appreciated that.

greedy52 replied:
this is a good call! I've added a `exec.LookPath(gitExecutable)` check with specific error message. `tsh git config` should be first command user uses when setting things up so proper message to ask user to install git is nice.

Tener commented on tool/tsh/common/git_config.go:
```diff
@@ -0,0 +1,178 @@
+
+func (c *gitConfigCommand) run(cf *CLIConf) error {
+	// Make sure we are in a Git dir.
+	err := execGitWithStdoutAndStderr(cf, io.Discard, io.Discard, "rev-parse", "--is-inside-work-tree")
+	if err != nil {
+		// This error message is a slight alternation of the original error
+		// message from the above command.
+		return trace.BadParameter("the current directory is not a Git repository (or any of the parent directories)")
+	}
```
I think that without `git` installed this will give incorrect error message ("not a Git repo" instead of "git not found").

Tener reviewed (commented):
Can you add some context to the implementation choice, i.e, shelling out to `git` binaries versus using `go-git`? There are some pros and cons to both and it would be beneficial to hear your reasoning; perhaps worth documenting it in the package itself, too?

On that note, I wonder if we should start splitting the tsh subcommands into their own packages.

We are using `git` command line interface as the API, which makes it hard to spot any potential mistakes. We don't have infrastructure for this yet, but having an end-to-end suite of tests (including a real github repo and docker image with `git`) would be amazing.

All those shell invocations makes me wonder about Windows. Can you check if `tsh.exe` works as intended? (My build command: `BUILDDIR=build-win FIDO2=no ARCH=amd64 OS=windows make build-win/tsh && mv build-win/tsh build-win/tsh.exe`)

greedy52 committed:
review comments

greedy52 commented:
> Can you add some context to the implementation choice, i.e, shelling out to `git` binaries versus using `go-git`? There are some pros and cons to both and it would be beneficial to hear your reasoning; perhaps worth documenting it in the package itself, too?

using `go-git` is an afterthought. By design, we use `git` under the hood. I've changed the command to a `TODO` to investigating the option to use `go-git`. but overall i feel `git` binary is more stable (in all platforms) than `go-git`). And for things like `tsh git clone --some-flag` (which yet to be implemented), it will be easier to pass extra args directly to `git` binary then re-implementing the flags through `go-git`.

> On that note, I wonder if we should start splitting the tsh subcommands into their own packages.

It still relies on `CLIConf`, so would require quite some refactoring to have its own package.

> We are using git command line interface as the API, which makes it hard to spot any potential mistakes. We don't have infrastructure for this yet, but having an end-to-end suite of tests (including a real github repo and docker image with git) would be amazing.

Added to TODO list (https://github.com/gravitational/teleport/issues/48762).

> All those shell invocations makes me wonder about Windows. Can you check if tsh.exe works as intended? (My build command: BUILDDIR=build-win FIDO2=no ARCH=amd64 OS=windows make build-win/tsh && mv build-win/tsh build-win/tsh.exe)

I think it should work without issue. I will give it a shot later and submit a separate issue to track it if it doesn't work.

greedy52 committed:
fix test

greedy52 committed:
fix ut for lookpath

greedy52 commented:
friendly ping @smallinsky @r0mant

greedy52 commented on tool/tsh/common/git_config.go:
```diff
@@ -0,0 +1,184 @@
+
+func (c *gitConfigCommand) run(cf *CLIConf) error {
+	// Make sure we are in a Git dir.
+	err := execGitWithStdoutAndStderr(cf, io.Discard, io.Discard, "rev-parse", "--is-inside-work-tree")
```
I don't see the `false` getting printed:
```
stevehuang@mac ~ $ git rev-parse --is-inside-work-tree
fatal: not a git repository (or any of the parent directories): .git
```

Same on my mac or my ec2 linux machine.

greedy52 committed:
Merge branch 'master' of github.com:gravitational/teleport into STeve/48762_ssh_tsh

greedy52 committed:
fix logger and update dependency version

greedy52 committed:
go mod tidy for integrations
